kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/DomainMetadataUtils.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (2024) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.delta.kernel.internal.util;
+
+import io.delta.kernel.data.ColumnVector;
+import io.delta.kernel.internal.DeltaErrors;
+import io.delta.kernel.internal.TableFeatures;
+import io.delta.kernel.internal.actions.DomainMetadata;
+import io.delta.kernel.internal.actions.Protocol;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class DomainMetadataUtils {
+
+  private DomainMetadataUtils() {
+    // Empty private constructor to prevent instantiation
+  }
+
+  /**
+   * Extracts a map of domain metadata from actions. There should be only one {@link DomainMetadata}
+   * per domain name. It throws a KernelException if duplicate domain metadata is found.
+   *
+   * @param domainMetadataActionVector A {@link ColumnVector} containing the domain metadata rows
+   * @return A map where the keys are domain names and the values are {@link DomainMetadata} objects
+   */
+  public static Map<String, DomainMetadata> extractDomainMetadataMap(
```

Comment:
This seems really similar to what we do in log replay. If we remove the duplicate check here can it be combined? If not can we at least clarify specifically when we use this method (during conflict resolution)?

Reply:
We could combine them if we decide to remove the duplicate check (during conflict resolution).

But we might need to adjust the method signature and usage. Currently, the method returns a map for each domainMetadata `ColumnVector`, but log replay requires building a single map after consuming all domainMetadata `ColumnVector`s.

We could create the map outside the method, and change the signature to something like
```
public static void fillDomainMetadataMap(ColumnVector domainMetadataActionVector, Map<String, DomainMetadata> mapToBeFilled)
```
This would allow it to accommodate both use cases.

Reply:
Yes I was thinking something like that. And during log replay we combine the maps. That plan sounds good to me.

Reply:
I've removed the duplicate check and adjust the signature to 
```
public static void populateDomainMetadataMap(
      ColumnVector domainMetadataActionVector, Map<String, DomainMetadata> domainMetadataMap)
```
And now it is used for both domain metadata conflict checking and log replay. Thanks for the suggestion!